extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -402,18 +535,14 @@ public void onSubscribe(Subscription s) {
             }
 
             @Override
-            public void onNext(ByteBuffer byteBuffer) {
-                if (!bufferReference.acquireIfAvailable()) {
+            public void onNext(final ByteBuffer byteBuffer) {
+                if (Request.super.get() == null) {
```

Comment:
If we make the proposed change, you should no longer call `super.get()` anywhere.

Also, we probably need more change here.
Using a duplicate buffer may keep a hard reference to the request buffer internally (source code suggests that this is only an issue with direct buffers, and we're using heap buffers, but that could change...).

I think we need to consider whether to continue using persistent duplicate buffers. We could instead just keep track of our offset (position) into the result buffer, and do:
```
                get().duplicate().position(offset).put(byteBuffer); // After checking result of get() for null
                offset += length;
```

Reply:
I think you still have a reachability issue here. You don't want the `Request` to directly or indirectly maintain a persistent hard reference to the buffer.

We're depending on implementation details for `duplicate()` by keeping the reference. We may also want to reconsider whether the read only buffer is the appropriate result for the future, rather than a `Boolean` or something.